core/trino-main/src/main/java/io/trino/sql/gen/ExpressionCompiler.java

```diff
@@ -138,20 +135,23 @@ private Supplier<PageProcessor> compilePageProcessor(
             List<PageProjection> pageProjections = pageProjectionSuppliers.stream()
                     .map(Supplier::get)
                     .collect(toImmutableList());
-            return new PageProcessor(filterEvaluator, pageProjections, initialBatchSize);
+            Optional<FilterEvaluator> dynamicFilterEvaluator = dynamicPageFilter
+                    .map(dynamicFilter -> dynamicFilter.createDynamicPageFilterEvaluator(columnarFilterCompiler))
+                    .map(Supplier::get);
```

Comment:
`Supplier::get` could be internal detail of `createDynamicPageFilterEvaluator`
OR
`dynamicFilter.createDynamicPageFilterEvaluator(columnarFilterCompiler)` should be obtained there:
```
Optional<Supplier<PageFilter>> finalFilterFunctionSupplier = filterFunctionSupplier;
Optional<Supplier<FilterEvaluator>> dynamicFilterSupplier = dynamicFilter.createDynamicPageFilterEvaluator(columnarFilterCompiler);
```

Reply:
See https://github.com/trinodb/trino/pull/22411#discussion_r1666444602 for explanation on why createDynamicPageFilterEvaluator this returns a Supplier. This is exactly the way existing compiled filter and project work as well.
We can't move `createDynamicPageFilterEvaluator` outside the supplier because this supplier is invoked per split and we want each split to get either updated or cached dynamic filter through `io.trino.sql.gen.columnar.DynamicPageFilter#createDynamicPageFilterEvaluator`

Reply:
exposing supplier looks odd when
```
                    .map(dynamicFilter -> dynamicFilter.createDynamicPageFilterEvaluator(columnarFilterCompiler))
                    .map(Supplier::get);
```
are called together. `get` could be internal to `createDynamicPageFilterEvaluator`. but it's a nit really

Reply:
Current way makes it easier to keep scope of `synchronized` limited to only where it's needed, creation of compiled class and running of constructor through the Supplier remains outside critical section.
This method also makes it easy to verify in tests that the recompilation is happening only when we expect it to happen, by checking for change in the Supplier.